### DISCUSSION TITLE
Question about optimizing URL representation for multiple select in Livewire

### DISCUSSION THREAD

**Discussion by plompd** [livewire/livewire#9066] [1. Help]
I’m working on a Livewire component with a filter containing multiple checkboxes. Initially, the selected values were displayed in the URL like this:

`?levels[0]=1&levels[1]=2`

However, I wanted to make the URL more user-friendly and readable by:

1. Displaying slugs instead of IDs: For example, `?levels=management,specialist` instead of IDs.
2. Removing the array notation (`levels[0]=1`): Instead, show a comma-separated list like `?levels=1,2`.
3. Avoiding repeated database queries: Ensure that no database queries are made during every `hydrate` and `dehydrate` lifecycle call in Livewire.
4. Maintaining checkbox functionality: The checkboxes remain bound to an array of selected values (`selectedLevels`), while the URL (`urlSelectedLevels`) handles slugs.

Here’s how I optimized the component:

**Step 1: Initializing Mapping in mount**
In the mount method, I load a mapping of IDs to their corresponding names and slugs, so it’s available throughout the component lifecycle:

```php
public array $levelMappings = [];

public function mount(): void
{
    $this->levelMappings = JobLevel::all()
        ->mapWithKeys(function ($level) {
            return [
                $level->id => [
                    'name' => $level->name,
                    'slug' => Str::slug($level->name),
                ],
            ];
        })
        ->toArray();
}
```

This results in a structure like this:

```php
[
    1 => ['name' => 'Management', 'slug' => 'management'],
    2 => ['name' => 'Specialist', 'slug' => 'specialist'],
    // ...
]
```

**Step 2: Updating the URL with Slugs in dehydrate**
During the dehydrate lifecycle hook, I use the mapping to generate slugs for the selected levels without querying the database:

```php
public function dehydrate(): void
{
    $this->urlSelectedLevels = implode(',', array_map(
        fn($id) => $this->levelMappings[$id]['slug'] ?? '',
        $this->selectedLevels
    ));
}
```

**Step 3: Converting Slugs Back to IDs in hydrate**
In the hydrate method, I convert slugs from the URL back to their corresponding IDs using the preloaded mapping:

```php
public function hydrate(): void
{
    if ($this->urlSelectedLevels) {
        $slugs = array_filter(explode(',', $this->urlSelectedLevels));

        $this->selectedLevels = array_keys(array_filter($this->levelMappings, function ($mapping) use ($slugs) {
            return in_array($mapping['slug'], $slugs);
        }));
    }
}
```

**Step 4: Resetting the Filters**
To reset the filters, I ensure both `selectedLevels` and `urlSelectedLevels` are cleared:

```php
public function resetLevels(): void
{
    $this->selectedLevels = [];
    $this->urlSelectedLevels = '';
    $this->resetPage();
}
```

**Results**
1. Optimized URL: The URL now shows slugs instead of IDs and uses a comma-separated format:

`?levels=management,specialist`

2. No Repeated Queries: All database queries are made only once during the mount lifecycle, and the mapping is reused in memory.
3. Checkbox Functionality Maintained: The checkboxes remain bound to selectedLevels, an array of IDs.
4. Scalability: The mapping can be extended easily if new levels are added, without additional queries.

**Initial Situation vs Current Solution**

- Initial URL:
`?levels[0]=1&levels[1]=2`

- Optimized URL:
`?levels=management,specialist`

So now my question is:
Is this the best way to handle URL adjustments for multiple checkboxes in Livewire? Are there any further optimizations or edge cases I should consider? Maybe there is something in Livewire I completely missed and this can be done much simpler?  I’d love to hear your feedback and suggestions! 

This is the dropdown:

<img width="185" alt="image" src="https://github.com/user-attachments/assets/0f6b34f5-dca8-4b75-88f6-4703d7bbc9f0">

**Comment by joshhanley** [comment]
@plompd looks good to me. We don't really have a better story for managing/ manipulating query string data, so what you have is what I would probably do 🙂
